[Fabric] Use pre allocated header on worker

daminakaTT opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/24850
https://github.com/tenstorrent/tt-metal/issues/24880

### Problem description
Developer have to do prepare buffer for header, then retrieve it from kernel which is not developer friendly

### What's changed
Pre-allocate on worker L1 memory `(64 bytes * (4 or 6 direction) * (2 riscs) * 2 (just convention))`. Now kernel just call `PACKET_HEADER_POOL_ALLOC` to get the address of header then use it as they want.

APC
https://github.com/tenstorrent/tt-metal/actions/runs/16515481880
TG
https://github.com/tenstorrent/tt-metal/actions/runs/16515480247
Benchmark
https://github.com/tenstorrent/tt-metal/actions/runs/16515478369
T3K
https://github.com/tenstorrent/tt-metal/actions/runs/16515477356
BH
https://github.com/tenstorrent/tt-metal/actions/runs/16515483601

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [x] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml) CI passes (if applicable)
- [x] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

daminakaTT committed:
use pre allocated header on worker

daminakaTT committed:
fix more tests

daminakaTT committed:
include header info

aagarwalTT commented on tt_metal/hw/inc/packet_header_pool.h:
```diff
@@ -0,0 +1,62 @@
+    FORCE_INLINE static void init() { current_offset_ = 0; }
+
+    FORCE_INLINE static volatile tt_l1_ptr PACKET_HEADER_TYPE* allocate_header() {
+        ASSERT(current_offset_ + HEADER_SIZE <= POOL_SIZE, "Insufficient space in packet header pool");
```
can we also clear out the region before allocating? this will reduce a lot of duplication on client side

yugaoTT commented on tt_metal/hw/inc/blackhole/dev_mem_map.h:
```diff
@@ -118,7 +118,21 @@
 
-#define MEM_MAP_END (MEM_TENSIX_FABRIC_CONNECTIONS_BASE + MEM_TENSIX_FABRIC_CONNECTIONS_SIZE)
+// Packet header pool sizing constants
+#define PACKET_HEADER_MAX_SIZE 64       //
+#define PACKET_HEADER_MAX_DIRECTIONS 6  // EAST, WEST, NORTH, SOUTH, UP, DOWN
+#define PACKET_HEADER_BUFFER_FACTOR 2   // Double buffering
```
is this arbitrary choice? and since the data in the packet header is not changing, it's not clear to me why we double buffering

yugaoTT commented on tests/ttnn/unit_tests/gtests/ccl/test_fabric_edm_common.hpp:
```diff
@@ -1185,7 +1163,7 @@ bool RunLineFabricTest(
         dest_is_dram,
         worker_buffer_index_semaphore_id,
-        packet_header_buffer_cb_id,
+        // packet_header_buffer_cb_id,
```
remove

yugaoTT commented on tt_metal/hw/inc/packet_header_pool.h:
```diff
@@ -0,0 +1,62 @@
+        return reinterpret_cast<volatile tt_l1_ptr PACKET_HEADER_TYPE*>(allocated_addr);
+    }
+
+    FORCE_INLINE static volatile tt_l1_ptr PACKET_HEADER_TYPE* allocate_headers(uint32_t count) {
+        ASSERT(count > 0, "Cannot allocate zero headers");
+
+        uint32_t total_size = count * HEADER_SIZE;
+        ASSERT(
+            current_offset_ + total_size <= POOL_SIZE,
+            "Insufficient space in packet header pool for {} headers",
+            count);
+
+        uint32_t allocated_addr = POOL_BASE + current_offset_;
+        add_and_wrap(total_size);
+
+        return reinterpret_cast<volatile tt_l1_ptr PACKET_HEADER_TYPE*>(allocated_addr);
+    }
```
this should return a list of ptrs back instead of one base address, so that user don't need to manually calculate their header adders?

daminakaTT committed:
fix compile time arg indexing

daminakaTT committed:
simplify header allocater

daminakaTT committed:
remove FACTOR

daminakaTT committed:
move file location. hang intentionally when allocation failed

daminakaTT commented on tt_metal/hw/inc/packet_header_pool.h:
```diff
@@ -0,0 +1,61 @@
+        ASSERT(current_offset_ + HEADER_SIZE <= POOL_SIZE, "Insufficient space in packet header pool");
+
+        uint32_t allocated_addr = POOL_BASE + current_offset_;
+        add_and_wrap(HEADER_SIZE);
```
Added `while (1) {}` for now

aagarwalTT commented on tt_metal/fabric/hw/inc/packet_header_pool.h:
```diff
@@ -0,0 +1,68 @@
+#include "core_config.h"
+
+static_assert(
+    proc_type == static_cast<uint8_t>(TensixProcessorTypes::DM0) ||
```
curious- we are only doing a static cast here and not down below (when arriving at the pool start offset), any reason for that?

daminakaTT committed:
fix for multi risc access pattern

daminakaTT committed:
remove cb allocation from test

daminakaTT committed:
Compress tensix_routing_l1_info_t size (#25382)

### Ticket
https://github.com/tenstorrent/tt-metal/issues/24850
https://github.com/tenstorrent/tt-metal/issues/24880

### Problem description
Tensix L1 is now fully utilized by newly added
`fabric_connection_info_t` and `tensix_fabric_connections_l1_info_t`.
This cause some workload to get OoM by Circular buffer allocation.

### What's changed
Compress `fabric_connection_info_t` by 3 bit encoding.
Reduced size from 2064 bytes -> 784 bytes. (1280 bytes reduction)

NOTE: Compressed version has additional overhead (2x slower) to read
values, but raw data is too huge (2048 bytes) to fit in L1 memory. Need
to evaluate once actual workloads are available

### Checklist
- [x] [All post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml)
CI passes
- [x] [Blackhole Post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml)
CI with demo tests passes (if applicable)
- [ ] [Model
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml)
CI passes (if applicable)
- [ ] [Device performance
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml)
CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml)
CI passes (if applicable) See [recommended dev
flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [x] [Galaxy
quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml)
CI passes (if applicable)
- [x] [TG demo tests, for
Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml)
CI passes, if applicable, because of current Llama work
- [x] (For runtime and ops writers) [T3000 unit
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml)
CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml)
CI passes (if applicable, since this is required for release)
- [ ] (For models and ops writers) [Single-card demo
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml)
CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

daminakaTT committed:
Merge branch 'main' into dev/daiki/packet_header_on_worker_L1

daminakaTT committed:
Merge branch 'main' into dev/daiki/packet_header_on_worker_L1

daminakaTT committed:
update ASSERT

ayerofieiev-tt commented on tt_metal/api/tt-metalium/fabric_host_interface.h:
```diff
@@ -64,6 +64,106 @@ enum eth_chan_directions : std::uint8_t {
 };
 
+// 3 bit expression
+enum compressed_routing_values : std::uint8_t {
```
as discussed, all this to me does not seem to belong to API
does not have to be made in this PR, but I think should be moved

SeanNijjar replied:
Thanks for bringing this up @ayerofieiev-tt - I agree this should not be in the fabric API file as the only consumer of it is metal itself (as opposed to say TT-NN or anything else built on top of Metalium)

ayerofieiev-tt commented on tt_metal/api/tt-metalium/fabric_host_interface.h:
```diff
@@ -64,6 +64,106 @@ enum eth_chan_directions : std::uint8_t {
+    inline void set_original_direction(std::uint16_t index, std::uint8_t original_direction) {
+        set_direction(index, compress_value(original_direction));
+    }
+} __attribute__((packed));
```
wonder if this attribute can go in front

daminakaTT committed:
follow comments

daminakaTT committed:
Merge branch 'main' into dev/daiki/packet_header_on_worker_L0

daminakaTT commented on tt_metal/api/tt-metalium/fabric_host_interface.h:
```diff
@@ -64,6 +64,106 @@ enum eth_chan_directions : std::uint8_t {
+};
+
+// Compressed routing table using 3 bits
+template <std::uint32_t ArraySize>
```
May I do it in next PR?
Move to `hostdevcommon` and split for host feature (.cpp) and device feature .h are complex and huge change

pgkeller replied:
hostdevcommon is deprecated and needs to go away.  fix is to split the code into device side (stays at the device), interface (in the device code but included in the host) and host side.  and all this stuff in the api needs to go away, these aren't apis

ayerofieiev-tt replied:
@daminakaTT , in scope of this PR.

CC @pgkeller

ayerofieiev-tt replied:
@daminakaTT said it is aligned over slack with @pgkeller to do it in the following PR.

daminakaTT committed:
Merge branch 'main' into dev/daiki/packet_header_on_worker_L0

ayerofieiev-tt reviewed (approved):
aligned next steps

tt-aho commented on tt_metal/fabric/hw/inc/packet_header_pool.h:
```diff
@@ -0,0 +1,60 @@
+
+uint32_t PacketHeaderPool::current_offset_ = PacketHeaderPool::risc_pool_start;
+
+#define PACKET_HEADER_POOL_ALLOC() PacketHeaderPool::allocate_header()
+#define PACKET_HEADER_POOL_RESET() PacketHeaderPool::init()
```
Why exactly did we need these define aliases?
Is PACKET_HEADER_POOL_RESET even used or can it be deleted? I don't see it being called anywhere?

daminakaTT committed:
Merge branch 'main' into dev/daiki/packet_header_on_worker_L1

daminakaTT committed:
remove macro

daminakaTT committed:
Merge branch 'main' into dev/daiki/packet_header_on_worker_L0
